### PR TITLE
feat!: rebuild remote cache with Files SDK, Nitro, and typed API routes

### LINKED ISSUE #1063: [Feature]: Full project revamp - no longer tied to Cloudflare
The release of [Files SDK](https://files-sdk.dev/) is a major milestone. This library should allow us to redhape our entire project to support effectively every provider out there.

What's more, [Nitro V3](https://nitro.build/) will ensure we can write a server that can be deployed on every major cloud platform.

This is the direction we should take this project in

### PR DESCRIPTION
**Pull request by AdiRishi**
## Overview

This is the major-version project revamp proposed in #1063. It replaces the bespoke Cloudflare-bound runtime and storage internals with Files SDK and Nitro, while deliberately keeping Cloudflare Workers and R2 as the simplest clone-and-deploy default.

Self-hosters can now select R2, KV, or S3 entirely through configuration, connect another Files SDK adapter through a typed source-level escape hatch, and build the same application for Cloudflare, Node, or another Nitro target.

The public cache routes are also rebuilt around Turborepo’s published OpenAPI contract. This closes gaps in the previous hand-written implementation—most notably batch artifact queries, complete artifact metadata, strict request validation, and structured errors—without adding another HTTP framework alongside Hono.

Closes #1063.

## Data flow

```text
Turbo client
  → Nitro entrypoint (or compatibility Worker facade)
  → normalize runtime bindings, environment, and waitUntil
  → Hono /v8 protocol router + Valibot boundary validation
  → Files SDK
  → R2 binding/credentials | KV binding | S3 | custom adapter
```

The scheduled retention task uses the same environment parsing and storage factory as request handling, so provider selection and retention behavior do not diverge between HTTP traffic and cleanup.

### Storage and deployment

- R2, KV, and S3 are first-class providers selected with `STORAGE_PROVIDER`; no source change is required.
- R2 supports either a native Workers binding or S3-compatible credentials. KV uses a project-owned Files SDK adapter because Files SDK does not ship a native Workers KV provider.
- Other Files SDK adapters can be connected through the typed `storage.config.ts` factory without complicating the default configuration path.
- Nitro now owns development, builds, deployment presets, and the `cache:delete-expired` task. Cloudflare remains the explicit default preset, with a Node build and `NITRO_PRESET` available for other targets.
- Retention uses Files SDK cursor pagination and batched deletion. KV additionally applies native per-key TTL; setting retention to `0` disables both mechanisms.

### Remote Cache API

- A canonical snapshot of Turborepo’s OpenAPI document is checked into the repository.
- Generation produces typed operation inputs/outputs plus a method/path manifest. TypeScript requires a handler for every published operation.
- Hono remains responsible for routing and streaming, with Valibot providing runtime validation at the HTTP boundary.
- All six published operations are covered, including batch artifact queries, `HEAD`, metadata-bearing uploads/downloads, status, and cache-event validation. Event persistence remains intentionally out of scope.
- Normal builds are network-free. `api:sync` is the explicit upstream refresh step, while `api:check` detects a non-canonical snapshot or stale generated contract.

## Compatibility and migration

The implementation preserves the existing `/v8/artifacts` paths, bearer authentication, `TURBO_TOKEN`, `R2_STORE` and `KV_STORE` bindings, legacy artifact metadata reads, and stored object keys when `STORAGE_PREFIX` is unset. Provider auto-detection also retains the legacy rule that KV wins when both Cloudflare bindings exist, and `BUCKET_OBJECT_EXPIRATION_HOURS` remains a fallback for the new retention setting.

The intentional major-version changes are:

- uploads now require `Content-Length`;
- artifact hashes must satisfy the published hexadecimal format;
- build and deployment output moves from Vite’s `dist/` to Nitro’s `.output/`;
- the old internal storage TypeScript APIs are replaced by Files SDK and `storage.config.ts`.

Changing providers or `STORAGE_PREFIX` does not migrate existing artifacts. A deployment moving to another provider should either accept a cold cache or separately copy object bodies, keys, and signed-artifact metadata. Likewise, rolling back after writing new artifacts may make those signed entries appear cold to the older release because its metadata reader predates the new cross-provider representation.

## Review notes

The most useful review order is the existing commit sequence: runtime/storage and migration first, the OpenAPI-backed protocol second, then the final boundary simplification.

Areas worth closer attention:

- `cloudflareKv` is the largest project-owned provider seam. It preserves legacy KV objects, enforces KV’s 25 MiB body and 1 KiB metadata limits, and buffers unknown-length streams within that bound.
- Native R2 requires a fixed-length stream marker. The storage helper restores it without buffering on Workers and uses a length-checked development fallback when `FixedLengthStream` is unavailable.
- Hono dispatches `HEAD` through the matching `GET` route, so the router explicitly chooses the `artifactExists` operation when the original request method is `HEAD`.
- Signed artifact metadata is stored through provider metadata. Adapters without that capability reject metadata-bearing uploads instead of silently producing unusable signed cache entries or creating sidecar objects.
- Nitro 3 task support is intentionally enabled despite its experimental status. The direct Wrangler entrypoint and cron remain as compatibility paths; the normal deployment uses Nitro’s generated Worker and schedule.
- The generated OpenAPI types and manifest enforce the compile-time contract, while the small handwritten Valibot layer remains the runtime boundary to avoid introducing a second routing/validation framework.

## Validation

All of the following passed locally:

```sh
pnpm api:check
pnpm cf-typegen --check
pnpm lint
pnpm typecheck
pnpm test
pnpm test:turbo
pnpm docs:build
pnpm build
pnpm build:node
```

The Vitest suite passes 104 tests across the protocol, provider selection, Workers KV adapter, runtime boundaries, retention, internal routes, and legacy compatibility. The Turbo smoke test starts the Nitro Cloudflare target and verifies a real signed cache miss followed by a remote cache hit that restores the artifact without rerunning the task.

### CODE REVIEW

File: worker-configuration.d.ts [hunk 1]
```diff
@@ -14,6 +14,12 @@ declare namespace Cloudflare {
 	interface Env extends __BaseEnv_Env {}
 }
 interface Env extends __BaseEnv_Env {}
+type StringifyValues<EnvType extends Record<string, unknown>> = {
+	[Binding in keyof EnvType]: EnvType[Binding] extends string ? EnvType[Binding] : string;
+};
+declare namespace NodeJS {
+	interface ProcessEnv extends StringifyValues<Pick<Cloudflare.Env, "STORAGE_PROVIDER" | "CACHE_RETENTION_HOURS" | "TURBO_TOKEN">> {}
```
[the remark this answers is not in the document]
Comment (AdiRishi) [reply worker-configuration.d.ts:21]: Validated against Wrangler 4.109 and Cloudflare's current type-generation/runtime docs. This block is generated, not project-authored: `wrangler types` deliberately defaults `--strict-vars` to `true`, and with our compatibility date `nodejs_compat` populates configured Worker variables into `process.env`. The portable Node path never assumes these individual properties exist; it spreads `process.env` into `RuntimeEnv`, whose fields are optional and runtime-validated. Hand-editing the declaration would immediately fail `pnpm cf-typegen --check`; `--strict-vars=false` would only discard useful literal types from the generated Cloudflare binding contract and still would not make the generated fields optional. I’m therefore leaving the reproducible generated output unchanged and resolving this as not applicable. References: https://developers.cloudflare.com/workers/wrangler/commands/workers/#types and https://developers.cloudflare.com/workers/configuration/environment-variables/#environment-variables-and-nodejs-compatibility

File: src/routes/internal.ts [hunk 2]
```diff
@@ -2,32 +2,21 @@ import { vValidator } from '@hono/valibot-validator';
...
   vValidator('json', v.object({ expireInHours: v.optional(v.number()) })),
   async (c) => {
     const { expireInHours } = c.req.valid('json');
-    await deleteOldCache({
-      ...c.env,
-      BUCKET_OBJECT_EXPIRATION_HOURS: expireInHours ?? c.env.BUCKET_OBJECT_EXPIRATION_HOURS,
-    });
+    await deleteOldCache(c.env.FILES, expireInHours ?? c.env.CACHE_RETENTION_HOURS);
     return c.json({ success: true });
   },
```
[the remark this answers is not in the document]
Comment (AdiRishi) [reply src/routes/internal.ts:21]: Confirmed: a negative override makes every dated object older than the computed future cutoff, so this was a real destructive edge case. Fixed in cf33456 by validating `expireInHours` with `v.minValue(0)`. The regression test sends `-1`, asserts HTTP 400, and verifies cleanup is never invoked. The full suite now passes 106 tests, along with lint, typecheck, API/type drift checks, the signed Turbo smoke test, and both deployment builds.
